### PR TITLE
Ontroller LEDs refresh timer

### PR DESCRIPTION
**Pull request by ghost**
Hi there,

I added a timer to refresh the LEDs every 2 seconds, without this the ontroller would reset it's LEDs after roughly 3s after the last WinUSB message.

By the way, i don't know if you saw this, but Dniel97 added MU3 LEDs support today to the hook, handling both io4 and 15093 boards! So Ontroller side LEDs is at arms reach!
https://gitea.tendokyu.moe/Dniel97/segatools/releases/tag/2024-05-12

Regards.

### CODE REVIEW

**Comment by jujuforce** [comment]
This is how https://gitea.tendokyu.moe/phantomlan/ontroller-hook is doing it but a delay added and not in a while loop. 

Here is a video of the issue before the proposed fix : 

https://github.com/SirusDoma/Mu3IO.NET/assets/9607300/119328b7-7c09-4ad0-a756-07ca0b9e9238

**Comment by SirusDoma** [comment]
Hi! Thanks for raising the PR!
I'm not really familiar with how the game deals with the LED and I don't have an ontroller myself so this is really helpful.

I have one concern about this change however, rather than using a `Timer`, I prefer to put this in `Poll` or other update method. I know `Poll` is not an ideal place, but I think I would rather avoid using a `Timer` (or any other independent timing mechanism), because they're simply aren't synchronized with game timing at all.

Let me hear what your thoughts

**Comment by SirusDoma** [comment]
Oh, and if you're to follow my suggestion, I would also like to hear the performance impact of writing the leds every time `Poll` gets called

**Comment by jujuforce** [comment]
Hi, it really seems overkill to do it every Poll, but it will work for sure and the impact will be non significant I think. The original Ontroller code, does it in a while loop like crazy, so ...
But maybe I can add a "lastLedRefreshTime" enforcing a 1 or 2 seconds wait in the poll method tho?
I can try that later this evening if you want?

Also, did you saw Dniel97 pre-release I was mentioning, maybe we could conform to the specs he describes in the mu3 segatools.ini :

```
; -----------------------------------------------------------------------------
; LED settings
; -----------------------------------------------------------------------------

[led15093]
; Enable emulation of the 15093-06 controlled lights, which handle the air tower 
; RGBs and the rear LED panel (billboard) on the cabinet.
enable=1

[led]
; Output billboard LED strip data to a named pipe called "\\.\pipe\ongeki_led"
cabLedOutputPipe=1
; Output billboard LED strip data to serial
cabLedOutputSerial=0

; Output slider LED data to the named pipe
controllerLedOutputPipe=1
; Output slider LED data to the serial port
controllerLedOutputSerial=0

; Serial port to send data to if using serial output. Default is COM5.
;serialPort=COM5
; Baud rate for serial data
;serialBaud=921600

; Data output a sequence of bytes, with JVS-like framing.
; Each "packet" starts with 0xE0 as a sync. To avoid E0 appearing elsewhere,
; 0xD0 is used as an escape character -- if you receive D0 in the output, ignore
; it and use the next sent byte plus one instead.
;
; After the sync is one byte for the board number that was updated, followed by
; the red, green and blue values for each LED.
;
; Board 0 has 61 LEDs:
;   [0]-[1]: left side button
;   [2]-[8]: left pillar lower LEDs
;   [9]-[17]: left pillar center LEDs
;   [18]-[24]: left pillar upper LEDs
;   [25]-[35]: billboard LEDs
;   [36]-[42]: right pillar upper LEDs
;   [43]-[51]: right pillar center LEDs
;   [52]-[58]: right pillar lower LEDs
;   [59]-[60]: right side button
;
; Board 1 has 6 LEDs:
;   [0]-[5]: 3 left and 3 right controller buttons
;
```

I'm not sure how the pipe thing works along with the new mu3_io_led_init and mu3_io_led_set_colors functions he mention in the release notes.

**Comment by SirusDoma** [comment]
> Hi, it really seems overkill to do it every Poll, but it will work for sure and the impact will be non significant I think.

No, it's not overkilled, in my opinion; the payload is small, and with USB, it is way too trivial. I believe the hardware itself **may** expect the data to be written continuously (don't quote me). The performance impact concern is pointed towards the comparison between calling it on Poll vs setting leds on demand (via `SetLeds`). And my main concern was more about the hook flow / structure, but I'm generally fine putting this in `Poll`.


> Also, did you saw Dniel97 pre-release I was mentioning, maybe we could conform to the specs he describes in the mu3 segatools.ini

I haven't, and probably I can't check this anytime soon as I don't have an ontroller and I have other things to do. I think the problem is that we need to figure out how to write the led data of side led of the ontroller (that is, if the left led is controllable at all)

**Comment by jujuforce** [comment]
> I believe the hardware itself **may** expect the data to be written continuously (don't quote me). 

It is, my testing on the real hardware confirm this, but not that often. The max timeout is shown in the video I posted earlier.

> The performance impact concern is pointed towards the comparison between calling it on Poll vs setting leds on demand (via `SetLeds`). And my main concern was more about the hook flow / structure, but I'm generally fine putting this in `Poll`.

Let's put it in poll and see how it goes then.

> > Also, did you saw Dniel97 pre-release I was mentioning, maybe we could conform to the specs he describes in the mu3 segatools.ini
> 
> I think the problem is that we need to figure out how to write the led data of side led of the ontroller (that is, if the left led is controllable at all)

Yes it is controllable, I analysed it, this is how the ontroller hook is doing it :

2024-05-12_23h00_53

2024-05-12_23h05_27

pBuf being the buffer sent by the Bepinex mod to the hook through some kind of shared memory. outBuffer being also modified through the setLeds call elsewhere.
In short, we have to get the side LED data from the new Dniel97 hook, format it the right way and append it to the USB payload. We may encounter the same issues as the ontroller hook + Bepinex mod tho, that is that setting all the LEDs at once can cause issues to the middle LEDs. We will need to mitigate those, this is what I was trying to do at first by patching their hook but this was too complicated and improving your project seemed the clever and open-source way of doing it.
Here is a video of the issues I am talking about :

https://github.com/SirusDoma/Mu3IO.NET/assets/9607300/1f147df3-a37b-4b44-b36d-17a153bcdc04

**Comment by jujuforce** [comment]
Here you go, it works great and the CPU usage is barely increased :)

**Comment by SirusDoma** [comment]
Hi, sorry for the late response.
I'm kind of busy this week but I'll get back to you once I'm free.
